### PR TITLE
setup/evaluate env context after setup steps context.

### PR DESCRIPTION
**Pull request by TingluoHuang**
fixing #308

### CODE REVIEW

File: src/Runner.Worker/StepsRunner.cs [hunk 1]
```diff
@@ -76,38 +76,38 @@ public async Task RunAsync(IExecutionContext jobContext)
                 // Start
                 step.ExecutionContext.Start();
 
-                // Populate env context for each step
-                Trace.Info("Initialize Env context for step");
+                // Initialize scope
+                if (InitializeScope(step, scopeInputs))
```
Comment (TingluoHuang) [inline src/Runner.Worker/StepsRunner.cs]: `steps` context get set up in `InitializeScope`

File: src/Runner.Worker/StepsRunner.cs [hunk 2]
```diff
@@ -76,38 +76,38 @@ public async Task RunAsync(IExecutionContext jobContext)
...
+                        var actionEnvironment = templateEvaluator.EvaluateStepEnvironment(actionStep.Action.Environment, step.ExecutionContext.ExpressionValues, VarUtil.EnvironmentVariableKeyComparer);
+                        foreach (var env in actionEnvironment)
+                        {
+                            envContext[env.Key] = new StringContextData(env.Value ?? string.Empty);
+                        }
                     }
-                }
 
-                // Initialize scope
-                if (InitializeScope(step, scopeInputs))
```
Comment (TingluoHuang) [inline src/Runner.Worker/StepsRunner.cs]: the change just moves this up.

File: src/Test/L0/Worker/StepsRunnerL0.cs [hunk 3]
```diff
@@ -477,7 +477,44 @@ public async Task PopulateEnvContextForEachStep()
...
+            {
+                // Arrange.
+                var env1 = new MappingToken(null, null, null);
+                env1.Add(new StringToken(null, null, null, "env1"), new StringToken(null, null, null, "100"));
+                var step1 = CreateStep(hc, TaskResult.Succeeded, "success()", env: env1, name: "foo");
+                _stepContext.SetOutput("", "foo", "test", "something", out string reference);
+
+                var env2 = new MappingToken(null, null, null);
+                env2.Add(new StringToken(null, null, null, "env1"), new StringToken(null, null, null, "1000"));
+                env2.Add(new StringToken(null, null, null, "env2"), new BasicExpressionToken(null, null, null, "steps.foo.outputs.test"));
```
Comment (TingluoHuang) [inline src/Test/L0/Worker/StepsRunnerL0.cs]: make sure `env` that reference `steps` context can be resolved.

File: src/Test/L0/Worker/StepsRunnerL0.cs [hunk 4]
```diff
@@ -508,8 +544,11 @@ private Mock<IActionRunner> CreateStep(TaskResult result, string condition, Bool
                         stepContext.Object.Result = r;
                     }
                 });
+            var trace = hc.GetTrace();
+            stepContext.Setup(x => x.Write(It.IsAny<string>(), It.IsAny<string>())).Callback((string tag, string message) => { trace.Info($"[{tag}]{message}"); });
             stepContext.Object.Result = result;
             step.Setup(x => x.ExecutionContext).Returns(stepContext.Object);
+            step.Setup(x => x.RunAsync()).Callback(() => { _stepContext.SetOutput(null, name, "test", "something", out string reference); }).Returns(Task.CompletedTask);
```
Comment (thboop) [inline src/Test/L0/Worker/StepsRunnerL0.cs]: Consider doing this in the test that needs it (PopulateEnvContextAfterSetupStepsContext), not the shared test setup code

Comment (TingluoHuang) [reply src/Test/L0/Worker/StepsRunnerL0.cs]: will do. :)

**Pull request merged**